Isosurfacing_3/include/CGAL/Isosurfacing_3/internal/Dual_contouring_internal.h

```diff
@@ -0,0 +1,351 @@
+// Copyright (c) 2022 INRIA Sophia-Antipolis (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Daniel Zint
+//                 Julian Stahl
+
+#ifndef CGAL_DUAL_CONTOURING_3_INTERNAL_DUAL_CONTOURING_3_H
+#define CGAL_DUAL_CONTOURING_3_INTERNAL_DUAL_CONTOURING_3_H
+
+#include <CGAL/license/Isosurfacing_3.h>
+
+#include <CGAL/Bbox_3.h>
+#include <CGAL/Origin.h>
+#include <CGAL/centroid.h>
+
+#include <Eigen/SVD>
+#include <array>
+#include <map>
+#include <mutex>
+#include <vector>
+
+namespace CGAL {
+namespace Isosurfacing {
+namespace internal {
+
+namespace Positioning {
+/**
+ * \ingroup PkgIsosurfacing3Ref
+ *
+ * \brief Computes the vertex position for a point in Dual Contouring using Quadric Error Metrics and the SVD pseudo
+ * inverse.
+ *
+ * \details
+ *
+ * \tparam use_bbox clamp vertex position to the bounding box of the cell
+ *
+ */
+template <bool use_bbox = false>
+class QEM_SVD {
+public:
+    /**
+     * \ingroup PkgIsosurfacing3Ref
+     *
+     * \brief Computes the vertex position for a point in Dual Contouring.
+     *
+     * \details
+     *
+     * \tparam Domain_ must be a model of `IsosurfacingDomainWithGradient`.
+     *
+     * \param domain the domain providing input data and its topology
+     * \param isovalue value of the isosurface
+     * \param cell the cell within the domain for which the vertex position ins computed
+     * \param point the point position of the vertex that belongs to that cell
+     *
+     * \return true, if the voxel intersects the isosurface
+     */
+    template <class Domain_>
+    bool position(const Domain_& domain, const typename Domain_::FT isovalue,
+                  const typename Domain_::Cell_descriptor& cell, typename Domain_::Point& point) const {
+        typedef typename Domain_::Point Point;
+        typedef typename Domain_::Geom_traits::Vector_3 Vector;
+        typedef typename Domain_::FT FT;
+
+        typename Domain_::Cell_vertices vertices = domain.cell_vertices(cell);
+
+        std::vector<Point> pos(vertices.size());
+        std::transform(vertices.begin(), vertices.end(), pos.begin(),
+                       [&](const auto& v) { return domain.position(v); });
```

Comment:
two needless copies, could use a transform iterator that dereferences on the fly in the centroid call